File: src/C3.Catalogue/Brands/BrandOperationResult.vb
Namespace Brands

    Public Enum BrandFailure
        None = 0
        NameRequired
        InvalidCode
        DuplicateCode
        NotFound
        ReferencedByModel
        StorageFailure
    End Enum

    Public NotInheritable Class BrandOperationResult

        Private Sub New()
        End Sub

        Public Property IsSuccess As Boolean
        Public Property Brand As Brand
        Public Property Failure As BrandFailure
        Public Property Message As String

        Public Shared Function Success(value As Brand) As BrandOperationResult
            Return New BrandOperationResult() With {
                .IsSuccess = True,
                .Brand = value,
                .Failure = BrandFailure.None,
                .Message = String.Empty
            }
        End Function

        Public Shared Function Failed(failure As BrandFailure, message As String) As BrandOperationResult
            Return New BrandOperationResult() With {
                .IsSuccess = False,
                .Failure = failure,
                .Message = message
            }
        End Function

    End Class

End Namespace

File: src/C3.Catalogue/Brands/BrandOperationResult.cs
namespace C3.Catalogue.Brands
{
    public enum BrandFailure
    {
        None = 0,
        NameRequired,
        InvalidCode,
        DuplicateCode,
        NotFound,
        ReferencedByModel,
        StorageFailure
    }

    public sealed class BrandOperationResult
    {
        private BrandOperationResult()
        {
        }

        public bool IsSuccess { get; set; }

        public Brand Brand { get; set; }

        public BrandFailure Failure { get; set; }

        public string Message { get; set; }

        public static BrandOperationResult Success(Brand value)
        {
            return new BrandOperationResult
            {
                IsSuccess = true,
                Brand = value,
                Failure = BrandFailure.None,
                Message = string.Empty
            };
        }

        public static BrandOperationResult Failed(BrandFailure failure, string message)
        {
            return new BrandOperationResult
            {
                IsSuccess = false,
                Failure = failure,
                Message = message
            };
        }
    }
}
